### ISSUE TITLE
How to get the training and validation losses/metrics per iteration

### ISSUE THREAD

**Issue by adkinsty** [sb-ai-lab/Py-Boost#6]
Is there any way to retrieve the training and/or validation losses and/or metrics for each iteration in a training run for later use? I would like to save these KPIs so that I can visualize them in a dashboard.

**Comment by adkinsty** [comment]
I see that the training losses per iteration can be obtained after training from `model.history`

**Comment by btbpanda** [comment]
@adkinsty thanks for your feedback! About your question:

1) `Loss` and `Metric` inside `Py-Boost` are very different entities (also it is so for each GBDT implementation). `Loss` should produce the gradeints and hessians for building a tree while `Metirc` produces the values to show in train log and used for early stopping. For example, when you use BCE to train, there is difference between `BCELoss` and `BCEMetric`, but `BCEMetric` is the default choice of metric for `BCELoss` if nothing else passed. So now we are discussing only `Metric`, since we are talking about model evaluation.

2) From the `model.history` list and train log you can obtain the results only for datasets, that are passed as the `eval_sets` to `.fit` method of both `GradientBoosting` and `SketchBoost`. So if you do like `model.fit(X, y, eval_sets = [{'X': X_test, 'y': y_test}])` you will see the `Metric` values on `X_test`. Unfortunately, you can not use multiple metrics to evaulate here. There is an option to pass multiple datasets to `eval_sets` list, then the last dataset will be used for early stopping, but from history and log you can obtain the results for all datasets. `Metric` values for train dataset is not saved since I believe it is not informative value, and it is not good practice to use train/valid gap as overfitting detector

3) However, if you want to monitor everything, it is definetly possible and below I show you multiple ways to do that from simple to efficient

#### Train as valid set
First option is to pass train datasets to `eval_sets` list. It will look like  

```py
model.fit(X, y, eval_sets = [{'X': X, 'y': y}, {'X': X_test, 'y': y_test}])
```
This is obviously unefficient, because GBDT will store on GPU and handle `X` dataset twice and you still will be able to evaluate only one metric

#### Predict after train
Second option is using `.predict_staged` method. It will returns the prediction result from all iterations after model is trained. On prediction results you can evaluate manually all the statistics you want

```py
preds = model.predict_staged(X_test, iterations=[100, 300, 500]) # preds shape is (n_iters, n_samples, n_outputs)
```
That is the better way especially if you don't need to evaluate every step but only from specific `iterations` list. But if you need to evaluate each step remember that `preds` array could be quite large

#### Callback
The last and actually the best way is writting your own `Callback` class where you will define what you want to compute and save on fly while model is trained
```py
from py_boost.callbacks.callback import Callback

class MultipleMetricsHistory(Callback):
   
    def __init__(self, metric_list):
       
        # assume list of callable(y_true, y_pred)
        self.metric_list = metric_list
        self.history = None
       
    def before_train(self, build_info):
       
        self.history = []
        # postprocess fn defined by loss function
        self.postprocess_fn = build_info['model'].loss.postprocess_output
       
    def after_iteration(self, build_info):
       
        iter_info = []
        # single train dataset
        train = build_info['data']['train']
        # lists of val datasets
        valid = build_info['data']['valid']
        # calc metric on train data
        y_pred = self.postprocess_fn(train['ensemble'])
       
        metrics = []
        for fn in self.metric_list:
            metrics.append(float(fn(train['target'], y_pred)))
        iter_info.append(metrics)
       
        # calc metrics on validation sets
        for y_true, y_pred in zip(valid['target'], valid['ensemble']):
            y_pred = self.postprocess_fn(y_pred)
            metrics = []
            for fn in self.metric_list:
                metrics.append(float(fn(y_true, y_pred)))
            iter_info.append(metrics)

        self.history.append(iter_info)
       
        return False
       
    def after_train(self, build_info):
       
        self.history = np.array(self.history)
       

history = MultipleMetricsHistory(...) # pass the metric list here
model = GradientBoosting(..., callbacks=[history, ])
model.fit(X, y, eval_sets=[{'X': X_val0, 'y': y_val0}, {'X': X_val1, 'y': y_val1}])

# should contain np.ndarray of size (n_iter, n_datasets, n_metrics)
# n_datasets = n_val_datasets + 1, fist one is train
evals = history.history
```

So you can try all of that and decide what is your best choice. If you still have any questions I will try to help you

**Comment by adkinsty** [comment]
@btbpanda Thanks for the detailed and helpful response!
